statsmodels/stats/tests/test_multi.py

```diff
@@ -390,7 +390,7 @@ def test_tukeyhsd():
     # check p-values (divergence of high values is expected)
     small_pvals_idx = [2, 5, 7, 9]
     assert_allclose(myres[8][small_pvals_idx], res[small_pvals_idx, 3],
```

Comment:
>  I have verified the data from R as what is correctly output, but now need to follow up to see why there is a difference.

I'm not sure I understand. Did you run the commands in R yourself, or did you see that it appeared to match the table on page 85 (pdf page 96) of [this book](http://www.ievbras.ru/ecostat/Kiril/R/Biblio_N/R_Eng/Bretz2011.pdf)?

Can you see what happens for the same data in `scipy.stats.tukey_hsd`?

Can you check against our `mpmath` version and produce a convergence plot like we did two summers ago?

Reply:
I tried to repeat this calculation with `tukey_hsd` but it seems to be doing a different job because the results are quite different. 

<detailed>

For that dataset, R prints:

```
   Loc Var    Y1    Y2
1   UF   M  81.0  80.7
2   UF   S 105.4  82.3
3   UF   V 119.7  80.4
4   UF   T 109.7  87.2
5   UF   P  98.3  84.2
6    W   M 146.6 100.4
7    W   S 142.0 115.5
8    W   V 150.7 112.2
9    W   T 191.5 147.7
10   W   P 145.7 108.1
11   M   M  82.3 103.1
12   M   S  77.3 105.1
13   M   V  78.4 116.5
14   M   T 131.3 139.9
15   M   P  89.6 129.6
16   C   M 119.8  98.9
17   C   S 121.4  61.9
18   C   V 124.0  96.2
19   C   T 140.8 125.5
20   C   P 124.8  75.7
21  GR   M  98.9  66.4
22  GR   S  89.0  49.9
23  GR   V  69.1  96.7
24  GR   T  89.3  61.9
25  GR   P 104.1  80.3
26   D   M  86.9  67.7
27   D   S  77.1  66.7
28   D   V  78.9  67.4
29   D   T 101.8  91.8
30   D   P  96.0  94.1
```

I loaded it into Python. 
```python3
import numpy as np
from scipy.stats import tukey_hsd

M = [81, 146.6, 82.3, 119.8, 98.9, 86.9, 80.7, 100.4, 103.1, 98.9, 66.4, 67.7]
P = [98.3, 145.7, 89.6, 124.8, 104.1, 96, 84.2, 108.1, 129.6, 75.7, 80.3, 94.1]
S = [105.4, 142, 77.3, 121.4, 89, 77.1, 82.3, 115.5, 105.1, 61.9, 49.9, 66.7]
T = [109.7, 191.5, 131.3, 140.8, 89.3, 101.8, 87.2, 147.7, 139.9, 125.5, 61.9, 91.8]
V = [119.7, 150.7, 78.4, 124, 69.1, 78.9, 80.4, 112.2, 116.5, 96.2, 96.7, 67.4]

all_data = np.array([M, P, S, T, V])
# I think in the book it averages
# between Y1 and Y2. That's what this does.
averages = (all_data[:, :6] + all_data[:, 6:])/2  
M, P, S, T, V = averages
m_r = averages.mean(axis=1)  # this is `m_r` in the `statsmodels` test above
res = tukey_hsd(M, P, S, T, V)
print(res)
```
This prints:
```
Tukey's HSD Pairwise Group Comparisons (95.0% Confidence Interval)
Comparison  Statistic  p-value  Lower CI  Upper CI
 (0 - 1)     -8.150     0.970   -46.558    30.258
 (0 - 2)      3.258     0.999   -35.149    41.666
 (0 - 3)    -23.808     0.385   -62.216    14.599
 (0 - 4)     -4.792     0.996   -43.199    33.616
 (1 - 0)      8.150     0.970   -30.258    46.558
 (1 - 2)     11.408     0.904   -26.999    49.816
 (1 - 3)    -15.658     0.753   -54.066    22.749
 (1 - 4)      3.358     0.999   -35.049    41.766
 (2 - 0)     -3.258     0.999   -41.666    35.149
 (2 - 1)    -11.408     0.904   -49.816    26.999
 (2 - 3)    -27.067     0.264   -65.474    11.341
 (2 - 4)     -8.050     0.971   -46.458    30.358
 (3 - 0)     23.808     0.385   -14.599    62.216
 (3 - 1)     15.658     0.753   -22.749    54.066
 (3 - 2)     27.067     0.264   -11.341    65.474
 (3 - 4)     19.017     0.600   -19.391    57.424
 (4 - 0)      4.792     0.996   -33.616    43.199
 (4 - 1)     -3.358     0.999   -41.766    35.049
 (4 - 2)      8.050     0.971   -30.358    46.458
 (4 - 3)    -19.017     0.600   -57.424    19.391
```
These clearly don't match the p-values in the book. I don't think `tukey_hsd` is wrong... it just seems to be doing something different. 

In the `statsmodels` code:
`m_r` is the means of each group.
6 seems to be the number of samples per group.
4 is # DFs (5 groups minus 1)
I don't know where 110.8 is coming from. 
Can you see what the code is doing with these things and match it against our implementation of TukeyHSD? Does it even look similar?

It would be nice if we could figure out what this test is doing, but it's not necessary. I guess just check the input to `studentized_range` that is leading to these p-values and compare against the mpmath implementation.

Reply:
I don't know what `statsmodels`' `tukeyhsd` is doing. `pairwise_tukeyhsd`,  on the other hand, matches our `tukey_hsd`. Append to the code above:
```python3
from statsmodels.stats.multicomp import pairwise_tukeyhsd
endog = averages.ravel()
groups = ['M']*6 + ['P']*6 + ['S']*6 + ['T']*6 + ['V']*6
res = pairwise_tukeyhsd(endog, groups, 0.05)
print(res)
```
prints:
````
 Multiple Comparison of Means - Tukey HSD, FWER=0.05 
=====================================================
group1 group2 meandiff p-adj   lower    upper  reject
-----------------------------------------------------
     M      P     8.15    0.9 -30.2595 46.5595  False
     M      S  -3.2583    0.9 -41.6678 35.1511  False
     M      T  23.8083 0.3858 -14.6011 62.2178  False
     M      V   4.7917    0.9 -33.6178 43.2011  False
     P      S -11.4083    0.9 -49.8178 27.0011  False
     P      T  15.6583 0.7289 -22.7511 54.0678  False
     P      V  -3.3583    0.9 -41.7678 35.0511  False
     S      T  27.0667 0.2641 -11.3428 65.4761  False
     S      V     8.05    0.9 -30.3595 46.4595  False
     T      V -19.0167  0.589 -57.4261 19.3928  False
-----------------------------------------------------
```
So I don't know what that test is doing and in the short term I don't care. I don't think it's even public code, right?

So for that test, just find out how `studentized_range` is being called and make sure it's accurate against mpmath. Thank you!

Reply:
I `statsmodels`' `tukeyhsd` (and the calculation in the book) are for a different test that uses the fact that the measurements were taken in five different locations. We are doing a regular Tukey HSD test that doesn't account for this sort of thing. So we don't really care what the calculation is doing.

Reply:
The author rounded `110.8254416667` to `110.8`.  That could cause the discrepancy in `p-values`. Maybe if you change that you won't have to change the tolerance.

```R
library(MASS)
options(digits=16)
data("immer", package = "MASS")
immer.aov <- aov((Y1 + Y2)/2 ~ Var + Loc, data = immer)
summary(immer.aov)
```
Prints:
```
            Df          Sum Sq         Mean Sq  F value     Pr(>F)    
Var          4  2654.986166667  663.7465416667  5.98912  0.0024526 ** 
Loc          5 10610.452000000 2122.0904000000 19.14804 5.2119e-07 ***
Residuals   20  2216.508833333  110.8254416667          
```

Reply:
Thanks for doing more legwork on this while I was out.

> The author rounded 110.8254416667 to 110.8.

This fixes the issue. With SciPy's studentized range we can actually INCREASE the tolerance here now that a more precise input is used. I will leave it as-is since without scipy.stats.studentized_range statsmodels' fallbacks don't allow for this increase precision.